File: cpp/62.Unique_Paths.cpp
// 记忆化搜索 + DP
class Solution {
public:
    int uniquePaths(int m, int n) {
    	vector<vector<int>> v;
    	for(int i = 0; i < m; ++i)	v.push_back(vector<int>(n, 0));
    	return solve(m, n, v);
    }

    int solve(int m, int n, vector<vector<int>>& v) {
    	if(m == 1 || n == 1)	return 1;
    	if(v[m-1][n-1] != 0)	return v[m-1][n-1];
    	v[m-1][n-1] = solve(m-1, n, v) + solve(m, n-1, v);
    	return v[m-1][n-1];
    }
};

File: go/62.Unique_Paths.go
// DP
func uniquePaths(m int, n int) int {
	if m == 1 || n == 1 {
		return 1
	}

	var dp [][]int
	for i := 0; i < m; i++ {
		dp = append(dp, make([]int, n))
	}
	for i := 0; i < n; i++ {
		dp[0][i] = 1
	}

	for i := 0; i < m; i++ {
		dp[i][0] = 1
	}

	// dp[i][j] = dp[i-1][j] + dp[i][j-1]
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			dp[i][j] = dp[i-1][j] + dp[i][j-1]
		}
	}

	return dp[m-1][n-1]
}